handlers/download.go

```diff
@@ -28,52 +28,45 @@ import (
 	"github.com/getsentry/sentry-go"
 )
 
-type source int
-
-//go:generate stringer -type=source
-const (
-	TwseDailyClose source = iota
-	TwseThreePrimary
-	TpexDailyClose
-)
-
 // download job to run in workerpool
 type downloadJob struct {
 	ctx       context.Context
 	date      string
 	respChan  chan *[]interface{}
 	rateLimit int
-	origin    source
+	origin    parser.Source
 }
 
 // batching download all the historical stock data
 func (h *handlerImpl) BatchingDownload(ctx context.Context, req *dto.DownloadRequest) {
 	respChan := make(chan *[]interface{})
-	defer close(respChan)
+	//defer close(respChan)
 
-	go generateJob(ctx, req.RewindLimit*-1, TwseDailyClose, req.RateLimit, respChan)
-	go generateJob(ctx, req.RewindLimit*-1, TpexDailyClose, req.RateLimit, respChan)
+	go generateJob(ctx, req.RewindLimit*-1, parser.TwseDailyClose, req.RateLimit, respChan)
+	go generateJob(ctx, req.RewindLimit*-1, parser.TpexDailyClose, req.RateLimit, respChan)
 
-	for {
-		select {
-		case <-ctx.Done():
-			log.Warn("=== terminate the downloading process ===")
-			return
-		case obj, ok := <-respChan:
-			if ok {
-				h.dataService.BatchCreateDailyClose(ctx, obj)
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
```

Comment:
This goroutine runs forever until os signal interrupted. should we know when to close the `respChan` or have a waitGroup to control the progress instead?